File: sdk/c/src/providers/ockito.c
/**
 * ockito -- https://ockito.com/web-api
 */

#include "tempmail_internal.h"
#include <ctype.h>

#define OCKITO_BASE "https://ockito.com/web-api"

static const char *ockito_json_headers[] = {"Accept: application/json", NULL};

static const char *ockito_post_json_headers[] = {
    "Accept: application/json", "Content-Type: application/json", NULL};

static char *ockito_trim_copy(const char *s) {
  if (!s)
    return tm_strdup("");
  while (*s && isspace((unsigned char)*s))
    s++;
  const char *end = s + strlen(s);
  while (end > s && isspace((unsigned char)end[-1]))
    end--;
  size_t len = (size_t)(end - s);
  char *out = (char *)malloc(len + 1);
  if (!out)
    return NULL;
  memcpy(out, s, len);
  out[len] = '\0';
  return out;
}

static char *ockito_urlencode(const char *s) {
  if (!s)
    return tm_strdup("");
  size_t len = strlen(s);
  char *out = (char *)malloc(len * 3 + 1);
  if (!out)
    return NULL;
  char *p = out;
  static const char hex[] = "0123456789ABCDEF";
  for (size_t i = 0; i < len; i++) {
    unsigned char c = (unsigned char)s[i];
    if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
        (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' ||
        c == '~') {
      *p++ = (char)c;
    } else {
      *p++ = '%';
      *p++ = hex[c >> 4];
      *p++ = hex[c & 15];
    }
  }
  *p = '\0';
  return out;
}

static char *ockito_pick_str(const cJSON *obj, const char **keys,
                             int key_count) {
  char *raw = tm_json_get_str(obj, keys, key_count);
  char *trimmed = ockito_trim_copy(raw);
  free(raw);
  return trimmed;
}

static void ockito_add_picked_str(cJSON *target, const char *name,
                                  const cJSON *source, const char **keys,
                                  int key_count) {
  char *value = ockito_pick_str(source, keys, key_count);
  cJSON_AddStringToObject(target, name, value ? value : "");
  free(value);
}

static cJSON *ockito_request_json(const char *method, const char *path,
                                  const char **headers, const char *body,
                                  long *status_out) {
  char url[1024];
  snprintf(url, sizeof(url), OCKITO_BASE "%s", path);
  tm_http_method_t m = strcmp(method, "POST") == 0 ? TM_HTTP_POST : TM_HTTP_GET;
  const char **use_headers = headers;
  if (!use_headers) {
    use_headers = ockito_json_headers;
  }
  tm_http_response_t *resp = tm_http_request(m, url, use_headers, body, 15);
  if (!resp)
    return NULL;
  if (status_out)
    *status_out = resp->status;
  cJSON *root = cJSON_Parse(resp->body);
  tm_http_response_free(resp);
  return root;
}

static char *ockito_pack_token(const char *access_token,
                               const char *refresh_token) {
  cJSON *root = cJSON_CreateObject();
  if (!root)
    return NULL;
  cJSON_AddStringToObject(root, "access_token",
                          access_token ? access_token : "");
  cJSON_AddStringToObject(root, "refresh_token",
                          refresh_token ? refresh_token : "");
  char *out = cJSON_PrintUnformatted(root);
  cJSON_Delete(root);
  return out;
}

static void ockito_unpack_token(const char *packed, char **access_token_out,
                                char **refresh_token_out) {
  *access_token_out = NULL;
  *refresh_token_out = NULL;
  const char *value = packed ? packed : "";
  while (*value && isspace((unsigned char)*value))
    value++;
  size_t len = strlen(value);
  while (len > 0 && isspace((unsigned char)value[len - 1]))
    len--;
  if (len == 0 || value[0] != '{')
    return;

  char *copy = (char *)malloc(len + 1);
  if (!copy)
    return;
  memcpy(copy, value, len);
  copy[len] = '\0';
  cJSON *root = cJSON_Parse(copy);
  free(copy);
  if (!root)
    return;

  *access_token_out =
      ockito_pick_str(root, (const char *[]){"access_token"}, 1);
  *refresh_token_out =
      ockito_pick_str(root, (const char *[]){"refresh_token"}, 1);
  cJSON_Delete(root);
  if (!*access_token_out)
    *access_token_out = tm_strdup("");
  if (!*refresh_token_out)
    *refresh_token_out = tm_strdup("");
}

static char *ockito_refresh_access_token(const char *refresh_token) {
  const char *headers[] = {"Accept: application/json",
                           "Content-Type: application/json", NULL};
  char auth[256];
  snprintf(auth, sizeof(auth), "Authorization: Bearer %s",
           refresh_token ? refresh_token : "");
  const char *req_headers[] = {headers[0], headers[1], auth, "X-PASSTHROUGH: Y",
                               NULL};
  long status = 0;
  cJSON *root =
      ockito_request_json("POST", "/grefresh", req_headers, "{}", &status);
  if (!root)
    return NULL;
  if (status < 200 || status >= 300) {
    cJSON_Delete(root);
    return NULL;
  }
  char *access_token =
      ockito_pick_str(root, (const char *[]){"access_token"}, 1);
  cJSON_Delete(root);
  return access_token;
}

static cJSON *ockito_fetch_bearer_json(const char *path, char **access_token,
                                       const char *refresh_token) {
  char auth[512];
  snprintf(auth, sizeof(auth), "Authorization: Bearer %s",
           *access_token ? *access_token : "");
  const char *req_headers[] = {"Accept: application/json", auth, NULL};
  long status = 0;
  cJSON *root = ockito_request_json("GET", path, req_headers, NULL, &status);
  if (!root)
    return NULL;
  if (status == 401) {
    cJSON_Delete(root);
    char *refreshed = ockito_refresh_access_token(refresh_token);
    if (!refreshed)
      return NULL;
    free(*access_token);
    *access_token = refreshed;
    snprintf(auth, sizeof(auth), "Authorization: Bearer %s", *access_token);
    const char *retry_headers[] = {"Accept: application/json", auth, NULL};
    root = ockito_request_json("GET", path, retry_headers, NULL, &status);
    if (!root)
      return NULL;
  }
  if (status < 200 || status >= 300) {
    cJSON_Delete(root);
    return NULL;
  }
  return root;
}

static cJSON *ockito_flatten_inbox_row(const cJSON *raw,
                                       const char *recipient) {
  cJSON *flat = cJSON_CreateObject();
  if (!flat)
    return NULL;
  ockito_add_picked_str(flat, "id", raw, (const char *[]){"uid"}, 1);
  ockito_add_picked_str(flat, "from", raw, (const char *[]){"sender"}, 1);
  cJSON_AddStringToObject(flat, "to", recipient ? recipient : "");
  ockito_add_picked_str(flat, "subject", raw, (const char *[]){"subject"}, 1);
  ockito_add_picked_str(flat, "text", raw, (const char *[]){"snippet"}, 1);
  ockito_add_picked_str(flat, "html", raw, (const char *[]){"html"}, 1);
  cJSON_AddItemToObject(
      flat, "date",
      cJSON_Duplicate(
          cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "timestamp"), 1));
  cJSON_AddBoolToObject(flat, "is_read", false);
  cJSON_AddItemToObject(flat, "attachments", cJSON_CreateArray());
  return flat;
}

static cJSON *ockito_flatten_message(const cJSON *raw, const char *recipient) {
  const cJSON *obj = cJSON_GetObjectItemCaseSensitive((cJSON *)raw, "obj");
  if (!cJSON_IsObject(obj))
    obj = raw;
  cJSON *flat = cJSON_CreateObject();
  if (!flat)
    return NULL;
  ockito_add_picked_str(flat, "id", raw, (const char *[]){"uid"}, 1);
  char *from = ockito_pick_str(obj,
                               (const char *[]){"sender_email", "SenderEmail",
                                                "from_", "From", "from",
                                                "sender_name", "SenderName"},
                               7);
  char *to = ockito_pick_str(obj, (const char *[]){"to", "To"}, 2);
  cJSON_AddStringToObject(flat, "from", from ? from : "");
  cJSON_AddStringToObject(flat, "to",
                          (to && to[0]) ? to : (recipient ? recipient : ""));
  ockito_add_picked_str(flat, "subject", obj,
                        (const char *[]){"subject", "Subject"}, 2);
  ockito_add_picked_str(flat, "text", obj, (const char *[]){"text"}, 1);
  ockito_add_picked_str(flat, "html", obj, (const char *[]){"html"}, 1);
  cJSON_AddItemToObject(
      flat, "date",
      cJSON_Duplicate(cJSON_GetObjectItemCaseSensitive((cJSON *)obj, "date"),
                      1));
  cJSON_AddBoolToObject(flat, "is_read", false);
  cJSON_AddItemToObject(flat, "attachments", cJSON_CreateArray());
  free(from);
  free(to);
  return flat;
}

tm_email_info_t *tm_provider_ockito_generate(void) {
  long status = 0;
  cJSON *login = ockito_request_json("POST", "/gtoken",
                                     ockito_post_json_headers, "{}", &status);
  if (!login)
    return NULL;
  if (status < 200 || status >= 300) {
    cJSON_Delete(login);
    return NULL;
  }

  char *access_token =
      ockito_pick_str(login, (const char *[]){"access_token"}, 1);
  char *refresh_token =
      ockito_pick_str(login, (const char *[]){"refresh_token"}, 1);
  if (!access_token || !access_token[0] || !refresh_token ||
      !refresh_token[0]) {
    free(access_token);
    free(refresh_token);
    cJSON_Delete(login);
    return NULL;
  }

  char auth[512];
  snprintf(auth, sizeof(auth), "Authorization: Bearer %s", access_token);
  const char *email_headers[] = {"Accept: application/json", auth, NULL};
  status = 0;
  cJSON *email_json =
      ockito_request_json("GET", "/email", email_headers, NULL, &status);
  if (!email_json) {
    free(access_token);
    free(refresh_token);
    cJSON_Delete(login);
    return NULL;
  }
  if (status < 200 || status >= 300) {
    free(access_token);
    free(refresh_token);
    cJSON_Delete(login);
    cJSON_Delete(email_json);
    return NULL;
  }

  char *email = NULL;
  cJSON *email_value = cJSON_GetObjectItemCaseSensitive(email_json, "email");
  if (cJSON_IsString(email_value) && email_value->valuestring) {
    email = ockito_trim_copy(email_value->valuestring);
  } else if (cJSON_IsObject(email_value)) {
    email = ockito_pick_str(email_value, (const char *[]){"email"}, 1);
  }
  if (!email || !strchr(email, '@')) {
    free(access_token);
    free(refresh_token);
    free(email);
    cJSON_Delete(login);
    cJSON_Delete(email_json);
    return NULL;
  }

  char *packed = ockito_pack_token(access_token, refresh_token);
  tm_email_info_t *info = tm_email_info_new();
  if (!info || !packed) {
    if (info)
      tm_free_email_info(info);
    free(access_token);
    free(refresh_token);
    free(email);
    free(packed);
    cJSON_Delete(login);
    cJSON_Delete(email_json);
    return NULL;
  }
  info->channel = CHANNEL_OCKITO;
  info->email = tm_strdup(email);
  info->token = packed;

  free(access_token);
  free(refresh_token);
  free(email);
  cJSON_Delete(login);
  cJSON_Delete(email_json);
  return info;
}

tm_email_t *tm_provider_ockito_get_emails(const char *token, const char *email,
                                          int *count) {
  *count = -1;
  char *access_token = NULL;
  char *refresh_token = NULL;
  ockito_unpack_token(token, &access_token, &refresh_token);
  if (!access_token || !access_token[0] || !refresh_token ||
      !refresh_token[0]) {
    free(access_token);
    free(refresh_token);
    return NULL;
  }
  if (!email || !email[0]) {
    free(access_token);
    free(refresh_token);
    return NULL;
  }

  char *enc_email = ockito_urlencode(email);
  if (!enc_email) {
    free(access_token);
    free(refresh_token);
    return NULL;
  }

  char path[1024];
  snprintf(path, sizeof(path), "/retrieve/%s/emails", enc_email);
  free(enc_email);

  cJSON *json = ockito_fetch_bearer_json(path, &access_token, refresh_token);
  if (!json) {
    free(access_token);
    free(refresh_token);
    return NULL;
  }

  cJSON *inbox = cJSON_GetObjectItemCaseSensitive(json, "inbox");
  int n = cJSON_IsArray(inbox) ? cJSON_GetArraySize(inbox) : 0;
  *count = n;
  if (n == 0) {
    cJSON_Delete(json);
    free(access_token);
    free(refresh_token);
    return NULL;
  }

  tm_email_t *out = tm_emails_new(n);
  if (!out) {
    cJSON_Delete(json);
    free(access_token);
    free(refresh_token);
    return NULL;
  }

  for (int i = 0; i < n; i++) {
    cJSON *row = cJSON_GetArrayItem(inbox, i);
    cJSON *flat = NULL;
    char *uid = ockito_pick_str(row, (const char *[]){"uid"}, 1);
    if (uid && uid[0]) {
      char *enc_uid = ockito_urlencode(uid);
      if (enc_uid) {
        char detail_path[1024];
        snprintf(detail_path, sizeof(detail_path), "/retrieve/%s/%s", email,
                 enc_uid);
        cJSON *detail =
            ockito_fetch_bearer_json(detail_path, &access_token, refresh_token);
        if (detail) {
          flat = ockito_flatten_message(detail, email);
          cJSON_Delete(detail);
        }
        free(enc_uid);
      }
    }
    if (!flat) {
      flat = ockito_flatten_inbox_row(row, email);
    }
    if (!flat) {
      cJSON *empty = cJSON_CreateObject();
      out[i] = tm_normalize_email(empty, email);
      cJSON_Delete(empty);
    } else {
      out[i] = tm_normalize_email(flat, email);
      cJSON_Delete(flat);
    }
    free(uid);
  }

  cJSON_Delete(json);
  free(access_token);
  free(refresh_token);
  return out;
}

File: sdk/rust/src/providers/ockito.rs
/*!
 * ockito 渠道实现
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::Value;

const BASE_URL: &str = "https://ockito.com/web-api";

fn any_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.trim().to_string(),
        Value::Number(n) => n.to_string(),
        Value::Bool(b) => b.to_string(),
        _ => String::new(),
    }
}

async fn request_json(
    method: &str,
    path: &str,
    headers: Option<Vec<(&str, String)>>,
    body: Option<Value>,
) -> Result<(u16, Value), String> {
    let builder = match method {
        "POST" => http_client().post(format!("{}{}", BASE_URL, path)),
        _ => http_client().get(format!("{}{}", BASE_URL, path)),
    };
    let mut req = builder.header("Accept", "application/json");
    if let Some(hs) = headers {
        for (k, v) in hs {
            req = req.header(k, v);
        }
    }
    if let Some(b) = body {
        req = req.header("Content-Type", "application/json").json(&b);
    }
    let resp = req
        .send()
        .await
        .map_err(|e| format!("ockito request failed: {}", e))?;
    let status = resp.status().as_u16();
    let text = resp
        .text()
        .await
        .map_err(|e| format!("ockito read response: {}", e))?;
    let json = if text.trim().is_empty() {
        Value::Object(Default::default())
    } else {
        serde_json::from_str(&text)
            .map_err(|_| format!("ockito invalid JSON: {} HTTP {}", path, status))?
    };
    Ok((status, json))
}

fn encode_token(access_token: &str, refresh_token: &str) -> String {
    serde_json::json!({
        "access_token": access_token,
        "refresh_token": refresh_token,
    })
    .to_string()
}

fn decode_token(token: &str) -> Result<(String, String), String> {
    let value = token.trim();
    if value.is_empty() || !value.starts_with('{') {
        return Err("ockito: invalid session token".into());
    }
    let data: Value =
        serde_json::from_str(value).map_err(|_| "ockito: invalid session token".to_string())?;
    let access_token = any_string(&data["access_token"]);
    let refresh_token = any_string(&data["refresh_token"]);
    if access_token.is_empty() || refresh_token.is_empty() {
        return Err("ockito: invalid session token".into());
    }
    Ok((access_token, refresh_token))
}

async fn refresh_access_token(refresh_token: &str) -> Result<String, String> {
    let (status, data) = request_json(
        "POST",
        "/grefresh",
        Some(vec![
            ("Authorization", format!("Bearer {}", refresh_token)),
            ("X-PASSTHROUGH", "Y".to_string()),
        ]),
        None,
    )
    .await?;
    if !(200..300).contains(&status) {
        return Err(format!("ockito grefresh http {}", status));
    }
    let access_token = any_string(&data["access_token"]);
    if access_token.is_empty() {
        return Err("ockito: invalid refresh response".into());
    }
    Ok(access_token)
}

async fn fetch_bearer_json(
    path: &str,
    access_token: &mut String,
    refresh_token: &str,
) -> Result<Value, String> {
    let (status, data) = request_json(
        "GET",
        path,
        Some(vec![("Authorization", format!("Bearer {}", access_token))]),
        None,
    )
    .await?;
    if status == 401 {
        *access_token = refresh_access_token(refresh_token).await?;
        let (_, retry) = request_json(
            "GET",
            path,
            Some(vec![("Authorization", format!("Bearer {}", access_token))]),
            None,
        )
        .await?;
        return Ok(retry);
    }
    if !(200..300).contains(&status) {
        return Err(format!("ockito http {}", status));
    }
    Ok(data)
}

fn flatten_inbox_row(raw: &Value, recipient: &str) -> Value {
    serde_json::json!({
        "id": any_string(&raw["uid"]),
        "from": any_string(&raw["sender"]),
        "to": recipient,
        "subject": any_string(&raw["subject"]),
        "text": any_string(&raw["snippet"]),
        "html": any_string(&raw["html"]),
        "date": raw.get("timestamp").cloned().unwrap_or(Value::Null),
        "is_read": false,
        "attachments": [],
    })
}

fn flatten_message(raw: &Value, recipient: &str) -> Value {
    let obj = raw
        .get("obj")
        .and_then(|v| v.as_object())
        .cloned()
        .map(Value::Object)
        .unwrap_or_else(|| raw.clone());
    let from = any_string(&obj["sender_email"]).chars().collect::<String>();
    let to = {
        let value = any_string(&obj["to"]);
        if value.is_empty() {
            recipient.to_string()
        } else {
            value
        }
    };
    let subject = any_string(&obj["subject"]);
    let text = any_string(&obj["text"]);
    let html = any_string(&obj["html"]);
    let date = obj.get("date").cloned().unwrap_or(Value::Null);
    let attachments: Vec<Value> = Vec::new();
    serde_json::json!({
        "id": any_string(&raw["uid"]),
        "from": from,
        "to": to,
        "subject": subject,
        "text": text,
        "html": html,
        "date": date,
        "is_read": false,
        "attachments": attachments,
    })
}

pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let (status, login) =
            request_json("POST", "/gtoken", None, Some(serde_json::json!({}))).await?;
        if !(200..300).contains(&status) {
            return Err(format!("ockito gtoken http {}", status));
        }
        let access_token = any_string(&login["access_token"]);
        let refresh_token = any_string(&login["refresh_token"]);
        if access_token.is_empty() || refresh_token.is_empty() {
            return Err("ockito: invalid token response".into());
        }

        let (status, email_json) = request_json(
            "GET",
            "/email",
            Some(vec![("Authorization", format!("Bearer {}", access_token))]),
            None,
        )
        .await?;
        if !(200..300).contains(&status) {
            return Err(format!("ockito email http {}", status));
        }
        let email_value = &email_json["email"];
        let mut email = if let Some(s) = email_value.as_str() {
            s.trim().to_string()
        } else if let Some(obj) = email_value.as_object() {
            any_string(&obj.get("email").cloned().unwrap_or(Value::Null))
        } else {
            any_string(email_value)
        };
        if email.is_empty() {
            email = any_string(&email_json["email"]);
        }
        if email.is_empty() || !email.contains('@') {
            return Err("ockito: invalid email response".into());
        }

        Ok(EmailInfo {
            channel: Channel::Ockito,
            email,
            token: Some(encode_token(&access_token, &refresh_token)),
            expires_at: None,
            created_at: None,
        })
    })
}

pub fn get_emails(token: &str, email: &str) -> Result<Vec<Email>, String> {
    let (mut access_token, refresh_token) = decode_token(token)?;
    let address = email.trim();
    if address.is_empty() {
        return Err("ockito: empty email".into());
    }

    block_on(async {
        let data = fetch_bearer_json(
            &format!("/retrieve/{}/emails", urlencoding::encode(address)),
            &mut access_token,
            &refresh_token,
        )
        .await?;
        let rows = data
            .get("inbox")
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();
        let mut out = Vec::with_capacity(rows.len());
        for row in &rows {
            let uid = row.get("uid").and_then(|v| v.as_str()).unwrap_or("").trim();
            if uid.is_empty() {
                out.push(normalize_email(&flatten_inbox_row(row, address), address));
                continue;
            }
            let detail = fetch_bearer_json(
                &format!(
                    "/retrieve/{}/{}",
                    urlencoding::encode(address),
                    urlencoding::encode(uid)
                ),
                &mut access_token,
                &refresh_token,
            )
            .await
            .unwrap_or_else(|_| row.clone());
            out.push(normalize_email(&flatten_message(&detail, address), address));
        }
        Ok(out)
    })
}
